### PR TITLE
Jobs crud

### PR DESCRIPTION
**Pull request by itsTony4dev**

### CODE REVIEW

File: packages/backend/src/api/job/jobService.ts [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+      const paginationInfo = {
+        currentPage: page,
+        pageSize: limit,
+        totalCount: totalCount,
+        totalPages: Math.ceil(totalCount / limit),
+      };
+
+      const successMessage = `Jobs fetched successfully. Page ${page} of ${paginationInfo.totalPages}. Total jobs: ${totalCount}`;
+
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobService.ts:53]: Log success messages.

Comment (Hkataya) [reply packages/backend/src/api/job/jobService.ts:53]: logger.info("[findJobs]: Fetched Jobs Successfully")

File: packages/backend/src/api/job/jobService.ts [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+      if (params.companyName) {
+        conditions.push(like(company.name, `%${params.companyName}%`));
+      }
+      const [totalCount, result] = await Promise.all([
+        jobRepository.findJobsCountAsync(conditions),
+        jobRepository.findJobsAsync(conditions, limit, offset),
+      ]);
+
+      if (result.length === 0) {
+        return new ServiceResponse(ResponseStatus.Failed, 'No jobs found', null, StatusCodes.NOT_FOUND);
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobService.ts:42]: Response not failed

File: packages/backend/src/api/job/jobService.ts [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+      return new ServiceResponse(ResponseStatus.Success, 'Job deleted', deleteJob, StatusCodes.OK);
+    } catch (ex) {
+      const errorMessage = `Error deleting job with id ${id}:, ${(ex as Error).message}`;
+      logger.error(errorMessage);
+      return new ServiceResponse(ResponseStatus.Failed, errorMessage, null, StatusCodes.INTERNAL_SERVER_ERROR);
+    }
+  },
+
+  updateJob: async (
+    id: string,
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobService.ts:92]: UpdateJobRequest object

File: packages/backend/src/api/job/jobService.ts [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    tags: string,
+    isApproved: boolean
+  ): Promise<ServiceResponse<Job | null>> => {
+    const job = await jobRepository.findJobByIdAsync(id);
+    if (!job) {
+      return new ServiceResponse(ResponseStatus.Failed, 'Job not found', null, StatusCodes.NOT_FOUND);
+    }
+    try {
+      const updateJob = await jobRepository.updateJobAsync(
+        id,
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobService.ts:112]: Send new job object

File: packages/backend/src/api/job/jobService.ts [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+      const findJob = await jobRepository.findJobByIdAsync(id);
+      if (!findJob) {
+        return new ServiceResponse(ResponseStatus.Failed, 'Job not found', null, StatusCodes.NOT_FOUND);
+      }
+
+      if (findJob.isApproved) {
+        return new ServiceResponse(ResponseStatus.Failed, 'Job already approved', null, StatusCodes.CONFLICT);
+      }
+      const approvedJob = await db.update(job).set({ isApproved: true }).where(eq(job.id, id)).returning();
+      return new ServiceResponse(
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobService.ts:181]: Add a success log

File: packages/backend/src/api/job/jobRouter.ts [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+
+  jobRegistry.registerPath({
+    method: 'get',
+    path: '/jobs',
+    tags: ['Job'],
+    responses: createApiResponse(z.array(JobSchema), 'Success'),
+  });
+
+  router.get('/', async (req: Request, res: Response) => {
+    const { pageIndex, pageSize, categoryId, levelId, companyName } = req.query;
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobRouter.ts:25]: Get for required jobs -> validation rules. Flow as follows:
Route: 
create GetJobRequest object
Validate object
if validation fails -> return 400
else -> send to service

**Review by Hkataya — CHANGES_REQUESTED** [review]
@itsTony4dev create requests folder for all the request objects ( GetJobsRequest, CreateJobRequest).

File: packages/backend/src/api/job/jobRepository.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    const updateJob = await db
+      .update(job)
+      .set({
+        ...updateJobRequest,
+        updatedAt: new Date(),
+      })
+      .where(eq(job.id, id))
+      .returning();
+    return JobSchema.parse(updateJob[0]);
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobRepository.ts:89]: could this be a nullpointerexception ? updateJob[0]

File: packages/backend/src/api/job/jobRepository.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  },
+
+  deleteJobByIdAsync: async (id: string): Promise<Job | null> => {
+    const result = await db.delete(job).where(eq(job.id, id));
+    if (result.length === 0) return null;
+    return JobSchema.parse(result[0]) || null;
+  },
+
+  updateJobAsync: async (id: string, updateJobRequest: JobRequest): Promise<Job> => {
+    console.log('update');
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobRepository.ts:79]: add logs instead of console.log

File: packages/backend/src/api/job/jobModel.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      }
+      return true;
+    },
+    {
+      message: 'External jobs must have an application link',
+      path: ['applicationLink'],
+    }
+  );
+
+export type JobRequest = z.infer<typeof createJobSchema>;
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobModel.ts:44]: move to requests folder

**Review by Hkataya — CHANGES_REQUESTED** [review]

File: packages/backend/src/api/job/jobRepository.ts (hunk 7 above)
Comment (itsTony4dev) [reply packages/backend/src/api/job/jobRepository.ts:89]: no because parse we are updating and if there are somthng wrong with the input the parse will throw an error so the catch of in the service will handle it

File: packages/backend/src/api/job/jobModel.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+export const JobSchema = z
+  .object({
+    id: z.string().uuid(),
+    title: z.string().min(2, 'Title cannot be empty').max(100, 'Title is too long'),
+    levelId: z.number().int().positive('Level ID must be a positive integer'),
+    categoryId: z.number().int().positive('Category ID must be a positive integer'),
+    typeId: z.number().int().positive('Type ID must be a positive integer'),
+    location: z.string().min(1, 'Location cannot be empty').max(100, 'Location is too long'),
+    description: z.string().min(10, 'Description is too short').max(5000, 'Description is too long'),
+    compensation: z.string().min(1, 'Compensation cannot be empty').max(100, 'Compensation description is too long'),
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobModel.ts:18]: this is non required -> check db model

File: packages/backend/src/api/job/jobModel.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    applicationLink: z.string().url('Application link must be a valid URL'),
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobModel.ts:19]: also nullable

File: packages/backend/src/api/job/jobModel.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    isExternal: z.boolean(),
+    companyId: z.string().uuid().nullable(),
+    createdAt: z.union([z.string().datetime(), z.date()]),
+    updatedAt: z.union([z.string().datetime(), z.date()]),
+    postedAt: z.union([z.string().datetime(), z.date(), z.null()]),
+    tags: z.string().max(500, 'Tags are too long'),
+    isApproved: z.boolean(),
+    categoryTitle: z.string().optional(),
+    levelTitle: z.string().optional(),
+    companyName: z.string().nullable().optional(),
```
Comment (Hkataya) [inline packages/backend/src/api/job/jobModel.ts:29]: Why do we have companyName ?

Comment (itsTony4dev) [reply packages/backend/src/api/job/jobModel.ts:29]: I join the tables and return the the names in addition so the frontend can display the name directly, i tho this would be better but if u want I can remmove it

File: packages/backend/src/server.ts [hunk 13]
```diff
@@ -28,6 +31,8 @@ app.use(requestLogger);
 // Routes
 app.use('/health-check', healthCheckRouter);
 app.use('/users', userRouter);
+app.use('/jobs', jobRouter);
+app.use('/resources', resourceRouter);
```
Comment (Hkataya) [inline packages/backend/src/server.ts:35]: resource Router should not be here. It should belong to another pr
